pkg/agent/multicluster/mc_route_controller.go

```diff
@@ -0,0 +1,465 @@
+				return err
+			}
+			c.installedCIImports.Delete(ciImport)
+		}
+	}
+	return nil
+}
+
+func (c *MCRouteController) containsCIImport(ciImports []*mcv1alpha1.ClusterInfoImport, ciImport *mcv1alpha1.ClusterInfoImport) bool {
+	for _, cii := range ciImports {
+		if cii.Name == ciImport.Name {
+			return true
+		}
+	}
+	return false
+}
+
+func (c *MCRouteController) cleanupFlows(gatewayName string) error {
+	if err := c.deleteMCNodeFlowsForGateway(gatewayName); err != nil {
+		return err
+	}
+	if err := c.ofClient.UninstallMulticlusterFlows(gatewayName); err != nil {
+		return err
+	}
+	klog.V(2).InfoS("Deleted flows for Gateway", "gateway", gatewayName)
+	return nil
+}
+
+func (c *MCRouteController) installFlowsOnGateway(gw *mcv1alpha1.Gateway) error {
+	// Install Gateway flows only when the Node is the active Gateway.
+	if err := c.ofClient.InstallMulticlusterClassifierFlows(gw.Name, config.DefaultTunOFPort, true); err != nil {
+		return err
+	}
+	return c.addMCFlowsForCIImps(gw, nil)
+}
+
+func (c *MCRouteController) installFlowsOnRegularNode(gw *mcv1alpha1.Gateway) error {
+	if err := c.ofClient.InstallMulticlusterClassifierFlows(gw.Name, config.DefaultTunOFPort, false); err != nil {
+		return err
+	}
+	return c.addMCFlowsForCIImps(gw, nil)
+}
+
+func (c *MCRouteController) addMCFlowsForCIImps(activeGW *mcv1alpha1.Gateway, ciImp *mcv1alpha1.ClusterInfoImport) error {
+	var err error
+	var allciImports []*mcv1alpha1.ClusterInfoImport
+	if ciImp != nil {
+		allciImports = append(allciImports, ciImp)
+	} else {
+		allciImports, err = c.ciImportLister.ClusterInfoImports(c.namespace).List(labels.Everything())
+		if err != nil {
+			return err
+		}
+		if len(allciImports) == 0 {
+			klog.V(2).InfoS("No remote ClusterInfo imported, do nothing")
+			return nil
+		}
+	}
+
+	var tunnelPeerIP net.IP
+	for _, ciImport := range allciImports {
+		tunnelPeerIP = getPeerGatewayIP(ciImport.Spec)
+		if tunnelPeerIP == nil {
+			klog.V(2).InfoS("No valid peer Gateway IP from ClusterInfoImport, skip updating openflow rules",
+				"clusterinfoimport", klog.KObj(ciImport), "node", c.nodeConfig.Name)
+			continue
+		}
+
+		klog.InfoS("Adding/Updating remote Gateway Node flows for Multi-cluster", "gateway", klog.KObj(activeGW),
+			"node", c.nodeConfig.Name, "peer", tunnelPeerIP)
+		allCIDRs := []string{ciImport.Spec.ServiceCIDR}
+		peerConfigs, err := generatePeerConfigs(allCIDRs, tunnelPeerIP)
+		if err != nil {
+			klog.ErrorS(err, "Parse error for serviceCIDR from remote cluster", "clusterinfoimport", ciImport.Name, "gateway", activeGW.Name)
+			return err
+		}
+		if activeGW.Name == c.nodeConfig.Name {
+			klog.V(2).InfoS("Adding/Updating flows to remote Gateway Node for Multi-cluster traffic", "clusterinfoimport", ciImport.Name, "cidrs", allCIDRs)
+			localGatewayIP := net.ParseIP(activeGW.GatewayIP)
+			if err := c.ofClient.InstallMulticlusterGatewayFlows(
+				activeGW.Name+ciImport.Spec.ClusterID,
```

Comment:
Do we need to encode GW name into the key? It might make cleanup complexer, considering we might separate GW and CIImport flow error state.

Reply:
Could we just use CIImport.Name, considering ClusterID in theory can change?

Reply:
Sorry, I didn't get your first comment's point, for now, we encode the GW name into the key which simply the clean up process for a CIImport because no need to check if current Node's Gateway or not again.
for the second one, I will change it to CIImport.Name, but if ClusterID is changed, it usually means ClusterSet is reset, all old MC resources should be clean up and new resources will be created. (this is a case we haven't support in MC controller)

Reply:
I meant why we need to add GW name to the flow cache key instead of just using the CIImport.Name? If we use only CIImport.Name, when we remove flows, we need not to know the GW name (which might be simpler depending on our implementation).